ch4: Request queue refactoring

raffenet opened the pull request:
## Pull Request Description

This PR builds on #5261 to further refactor the ch4 request queues. I am opening it for comments while we wait for confirmation that removing the per-comm queues is acceptable.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

raffenet commented:
test:mpich/ch4/most

raffenet commented:
Additional note that for https://github.com/pmodels/mpich/pull/5280/commits/fad5ab5d5f6b19a2b61d68cee5091c81ab4104a4: moving the prev/next pointers into the main `MPIR_Request` struct increases the size of that struct from 480->496 bytes. I think in this is justified because it eliminates a dereference during each list traversal operation, and it doesn't bump the request over an additional cache line threshold on x86_64.

raffenet commented:
> Additional note that for [fad5ab5](https://github.com/pmodels/mpich/commit/fad5ab5d5f6b19a2b61d68cee5091c81ab4104a4): moving the prev/next pointers into the main `MPIR_Request` struct increases the size of that struct from 480->496 bytes. I think in this is justified because it eliminates a dereference during each list traversal operation, and it doesn't bump the request over an additional cache line threshold on x86_64.

Some shm performance measurements would probably be useful to verify these thoughts. I can work on those next week.

raffenet commented:
> > Additional note that for [fad5ab5](https://github.com/pmodels/mpich/commit/fad5ab5d5f6b19a2b61d68cee5091c81ab4104a4): moving the prev/next pointers into the main `MPIR_Request` struct increases the size of that struct from 480->496 bytes. I think in this is justified because it eliminates a dereference during each list traversal operation, and it doesn't bump the request over an additional cache line threshold on x86_64.
> 
> Some shm performance measurements would probably be useful to verify these thoughts. I can work on those next week.

Turns out I was wrong on this one. The extended request struct was only used for the out-of-order AM completion list, not match lists. I removed the modifications to the out-of-order completion list in this version because it was causing a deadlock that I couldn't not track down the root cause of...

raffenet commented:
test:mpich/ch4/most

wesbland reviewed (approved):
Seems reasonable to me. 👍

raffenet commented:
Cleaned up the history a bit in this version and fully converted partitioned match queues to use the generic utilities. Also added back unexpected queue events, which were never part of the global queue configuration 😅.
test:mpich/ch4/most

raffenet commented:
And not to contradict myself yet again, but this *does* move the list pointers into the main `MPIR_Request` struct and out of the "extended" struct. shm performance numbers coming soon...

raffenet commented:
This is a 2 process shared memory benchmark from JLSE Gomez nodes. I don't see a discernible difference between the 2 runs. Perhaps a more targeted match performance benchmark could shed more light. I'm not familiar with any such benchmark off the top of my head, but will give a look for the future. In any case, I think this should be safe to merge.
### Queue pointers in extended request struct
```
# OSU MPI Multiple Bandwidth / Message Rate Test
# [ pairs: 1 ] [ window size: 64 ]
# Size                    MB/s          Messages/s
1                         3.39          3389650.97
2                         7.12          3558639.71
4                        13.80          3451249.22
8                        28.53          3565771.49
16                       53.30          3331303.32
32                      114.97          3592962.29
64                      220.10          3439112.92
128                     426.57          3332611.27
256                     800.84          3128267.27
512                    1465.29          2861902.05
1024                   2720.12          2656366.61
2048                   4254.20          2077244.28
4096                   6392.22          1560601.07
8192                   9452.79          1153905.37
16384                  9907.33           604695.17
32768                 10806.95           329802.05
65536                 11272.23           172000.61
131072                10596.31            80843.41
262144                 9646.70            36799.24
524288                 9378.60            17888.25
1048576                9472.36             9033.55
2097152                9526.98             4542.82
4194304                9538.28             2274.10
```
### Queue pointers in MPIR_Request struct
```
# OSU MPI Multiple Bandwidth / Message Rate Test
# [ pairs: 1 ] [ window size: 64 ]
# Size                    MB/s          Messages/s
1                         3.85          3847999.22
2                         7.12          3559587.78
4                        14.84          3709224.61
8                        29.27          3658407.62
16                       59.65          3727841.91
32                      113.83          3557178.02
64                      226.47          3538570.42
128                     408.50          3191380.08
256                     818.05          3195521.48
512                    1517.70          2964259.83
1024                   2703.24          2639880.68
2048                   4536.28          2214978.93
4096                   6504.31          1587966.78
8192                   9680.79          1181736.70
16384                 10114.34           617330.10
32768                 10795.10           329440.36
65536                 11378.59           173623.52
131072                10797.95            82381.82
262144                 9667.00            36876.67
524288                 9406.32            17941.14
1048576                9466.98             9028.42
2097152                9497.25             4528.64
4194304                9516.51             2268.91
```

raffenet commented:
test:mpich/ch4/most

raffenet commented:
#5261 has been approved by downstream folks, and this PR that builds on it should be ready for review. I think we can skip the other one and focus our efforts here.

raffenet commented:
test:mpich/ch4/most

raffenet commented:
Fixed a couple of whitespace issues I noticed when going back over the commits. Re-testing...
test:mpich/ch4/most

raffenet commented:
test:mpich/ch4/most

hzhou reviewed (commented):
Looks good. Just one comment

raffenet commented on src/mpid/ch4/generic/am/mpidig_am_recv.h:
```diff
@@ -137,7 +137,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_do_irecv(void *buf, MPI_Aint count, MPI_Data
 
     root_comm = MPIDIG_context_id_to_comm(context_id);
-    unexp_req = MPIDIG_dequeue_unexp(rank, tag, context_id, &MPIDI_global.unexp_list);
+    unexp_req = MPIDIG_find_unexp(rank, tag, context_id, &MPIDI_global.unexp_list, true);
```
Yeah, I agree the `true|false` params are not easy to decipher. I'll incorporate your suggestion.

raffenet commented:
I ended up squashing two of the commits that made the queue functions generic in this version. Retesting.
test:mpich/ch4/most

hzhou reviewed (approved):
LGTM
